app/scripts/services/Message.js

```diff
@@ -0,0 +1,18 @@
+(function() {
+    function Message($firebaseArray) {
+        
+        var Message = {};
+        var ref = firebase.database().ref().child('messages');
+        var messages = $firebaseArray(ref);
+        
+        Message.getByRoomId = function(roomId) {
+            return ref.orderByChild('roomId').equalTo(roomId);
```

Comment:
Hey @jasonli12, you're getting a pretty strange error in the console that includes really technical Angular-speak like:

```
Can't interpolate: 
            {{homepage.activeRoomMessages}}
        
TypeError: JSON.stringify cannot serialize cyclic structures
```

That's a really complicated way of saying that Angular can't convert whatever `homepage.activeRoomMessages` into a JSON string that can be displayed on the page.

In this function here, you're sorting a Firebase reference and filtering out the messages by the room ID, but... it's not getting put into a format that Angular can display. Try wrapping this in a `$firebaseArray` function to convert it into an array. Then Angular will be able to display it. Basically it's a weirdly formatted Firebase object when it comes straight from Firebase, and we need to convert it into a clean array. Try console logging this line instead of returning it to see the Firebase object, and then console log it after you add `$firebaseArray` to see the difference:

```javascript
return $firebaseArray(ref.orderByChild('roomId').equalTo(roomId));
```